### ISSUE TITLE
Sphinx >=1.7.0 breaks help on Windows with a working dir on a drive different from the Python install drive

### ISSUE THREAD

**Issue by VanitarNordic** [spyder-ide/spyder#6549]
I faced this error while I was coding (typing) in the editor:

after closing it, the Spyder continues to its normal operation, I did not see any after effect.

Windows 10 - Python 3.6.4. The `Spyder 3.2.7` was installed using Anaconda

![2018-02-28_11-48-56](https://user-images.githubusercontent.com/18719591/36777332-5fb792d6-1c7e-11e8-9913-0ff5c7be1211.jpg)

**Comment by DanielAbdelNour** [comment]
I had to downgrade Sphinx to resolve this. It affected Spyder's help section rendering in my case.
This can be recreated by following the menu path Help > Tutorial.

**Comment by CAM-Gerlach** [comment]
Thanks for reporting. I was unable to reproduce this on Windows 8.1 with the same Python and Spyder versions when I selected ``Help > Tutorial``, and haven't seen it yet. Do you have Spyder/Anaconda installed in a different drive than your boot disk? That's what the error messages seems to suggest...

**Comment by VanitarNordic** [comment]
@CAM-Gerlach 
You're welcome. No I have the Anaconda on the boot drive which is C. This message appeared only after upgrading the Sphinx. in my case this message appeared when I start a new instance of the Spyder and start coding using autocomplete.

**Comment by CAM-Gerlach** [comment]
> This message appeared only after upgrading the Sphinx

Upgrading from which to which version?

> in my case this message appeared when I start a new instance of the Spyder and start coding using autocomplete.

So it only appears once per session, and never again? Is it consistant? WHat do you mean, "using autocomplete"? What sequence(s) of actions exactly triggers?

It seems the proximate cause might be related to [this Python/windows bug](https://bugs.python.org/issue7195), but I'm not quite sure yet what could actually be triggering it in your case. Is any of your code on anther drive?

**Comment by VanitarNordic** [comment]
as the picture says, upgrading the Sphinx to 1.7.1. 

Let me add some information and update the previous post. @DanielAbdelNour  is also correct. 

I captured a video from the screen to how reproduce it. This proves you can reproduce the error more than once and it is consistent and it is related to the help.

![2018-03-02_2-36-06](https://user-images.githubusercontent.com/18719591/36874806-8c520ea8-1dc2-11e8-9c4f-dbbb16c97ecd.gif)

**Comment by CAM-Gerlach** [comment]
> as the picture says, upgrading the Sphinx to 1.7.1.

As my question says, from which version?

> I captured a video from the screen to how reproduce it.

Thanks.

> This proves you can reproduce the error more than once and it is consistent and it is related to the help.

More than once _in the same session_, or just generally? 

Regardless, there seems to be something machine specific going on here, as again, I'm unable to reproduce it on my system with Python 3.6.4, Spyder 3.2.7, Sphinx 1.7.1 and Windows 8.1. Please post the full system information that was requested in the issue template (versions of Qt, PyQt, and dependencies) to help us narrow this down. 

@DanielAbdelNour Can you do the same? Also, downgrade sphinx to what version? 1.7.0? 1.6.6? Something else?

Thanks.

**Comment by VanitarNordic** [comment]
@CAM-Gerlach 

Sorry I did not get that question. The answer is that I remember that I upgraded the Sphinx two times. once from something to 1.7,0 and the last from 1.7.0 to 1.7.1. so what version it had was just before 1.7.0. Besides, this error was producing with 1.7.0 also. I'll put the whole Conda environment information.

**Comment by VanitarNordic** [comment]
```
# Name                    Version                   Build  Channel
absl-py                   0.1.9                     <pip>
alabaster                 0.7.10           py36hcd07829_0
apipkg                    1.4                       <pip>
asn1crypto                0.24.0                   py36_0
astroid                   1.6.1                    py36_0
attrs                     17.4.0                    <pip>
autopep8                  1.3.4                     <pip>
babel                     2.5.3                    py36_0
backports                 1.0              py36h81696a8_1
backports.weakref         1.0.post1        py36hf2695fa_0
bleach                    2.1.2                    py36_0
ca-certificates           2017.08.26           h94faf87_0
certifi                   2018.1.18                py36_0
cffi                      1.11.4           py36hfa6e2cd_0
chardet                   3.0.4            py36h420ce6e_1
cloudpickle               0.5.2                    py36_1
colorama                  0.3.9            py36h029ae33_0
contextlib2               0.5.5                     <pip>
cryptography              2.1.4            py36he1d7878_0
cycler                    0.10.0                    <pip>
decorator                 4.2.1                    py36_0
docutils                  0.14             py36h6012d8f_0
entrypoints               0.2.3            py36hfd66bb0_2
execnet                   1.5.0                     <pip>
futures                   3.1.1                     <pip>
glob2                     0.6                       <pip>
h5py                      2.7.1                     <pip>
html5lib                  1.0.1            py36h047fa9f_0
icc_rt                    2017.0.4             h97af966_0
icu                       58.2                 ha66f8fd_1
idna                      2.6              py36h148d497_1
imagesize                 1.0.0                    py36_0
imgaug                    0.2.5                     <pip>
intel-openmp              2018.0.0             hd92c6cd_8
ipykernel                 4.8.2                    py36_0
ipython                   6.2.1            py36h9cf0123_1
ipython_genutils          0.2.0            py36h3c5d0ee_0
isort                     4.3.4                    py36_0
jedi                      0.11.1                   py36_0
jinja2                    2.10             py36h292fed1_0
jpeg                      9b                   hb83a4c4_2
jsonschema                2.6.0            py36h7636477_0
jupyter_client            5.2.2                    py36_0
jupyter_core              4.4.0            py36h56e9d50_0
Keras                     2.1.4                     <pip>
lazy-object-proxy         1.3.1            py36hd1c21d2_0
libpng                    1.6.34               h79bbb47_0
libprotobuf               3.4.1                h3dba5dd_0
llvmlite                  0.21.0                    <pip>
lxml                      4.1.1                     <pip>
Markdown                  2.6.11                    <pip>
markupsafe                1.0              py36h0e26971_1
matplotlib                2.1.0                     <pip>
mccabe                    0.6.1            py36hb41005a_1
mistune                   0.8.3                    py36_0
mkl                       2018.0.1             h2108138_4
mock                      2.0.0                     <pip>
nbconvert                 5.3.1            py36h8dc0fde_0
nbformat                  4.4.0            py36h3a5bc1b_0
networkx                  2.1                       <pip>
numba                     0.36.2                    <pip>
numpy                     1.14.1           py36h4a99626_1
numpydoc                  0.7.0            py36ha25429e_0
opencv-python             3.4.0.12                  <pip>
openssl                   1.0.2n               h74b6da3_0
packaging                 16.8             py36ha0986f6_1
pandas                    0.22.0           py36h6538335_0
pandoc                    1.19.2.1             hb2460c7_1
pandocfilters             1.4.2            py36h3ef6317_1
parso                     0.1.1            py36hae3edee_0
path.py                   11.0                      <pip>
pathlib                   1.0.1                     <pip>
pbr                       3.1.1                     <pip>
phased-lstm-keras         1.0.2                     <pip>
pickleshare               0.7.4            py36h9de030f_0
Pillow                    5.0.0                     <pip>
pip                       9.0.1                    py36_5
pluggy                    0.6.0                     <pip>
prompt_toolkit            1.0.15           py36h60b8f86_0
protobuf                  3.5.1                     <pip>
psutil                    5.4.3            py36hfa6e2cd_0
py                        1.5.2                     <pip>
pycodestyle               2.3.1            py36h7cc55cd_0
pycparser                 2.18             py36hd053e01_1
pyflakes                  1.6.0            py36h0b975d6_0
pygments                  2.2.0            py36hb010967_0
pylint                    1.8.2                    py36_0
pyopenssl                 17.5.0           py36h5b7d817_0
pyparsing                 2.2.0            py36h785a196_1
pyparsing                 2.2.0                     <pip>
pyqt                      5.6.0            py36hb5ed885_5
pysocks                   1.6.8                    py36_0
pytest                    3.4.0                     <pip>
pytest-shutil             1.2.11                    <pip>
python                    3.6.4                h6538335_1
python-dateutil           2.6.1            py36h509ddcb_1
pytz                      2018.3                   py36_0
PyWavelets                0.5.2                     <pip>
PyYAML                    3.12                      <pip>
pyzmq                     16.0.3           py36he714bf5_0
qt                        5.6.2           vc14h6f8c307_12  [vc14]
qtawesome                 0.4.4            py36h5aa48f6_0
qtconsole                 4.3.1            py36h99a29a9_0
qtpy                      1.3.1            py36hb8717c5_0
requests                  2.18.4           py36h4371aae_1
rope                      0.10.7           py36had63a69_0
scikit-image              0.13.1                    <pip>
scikit-learn              0.19.1                    <pip>
scipy                     1.0.0                     <pip>
setuptools                38.5.1                   py36_0
simplegeneric             0.8.1                    py36_2
sip                       4.18.1           py36h9c25514_2
six                       1.11.0           py36h4db2310_1
snowballstemmer           1.2.1            py36h763602f_0
sphinx                    1.7.1                    py36_0
sphinxcontrib             1.0              py36hbbac3d2_1
sphinxcontrib-websupport  1.0.1            py36hb5e5916_1
spyder                    3.2.7                    py36_0
sqlite                    3.22.0               h9d3ae62_0
tensorflow                1.5.0                     <pip>
tensorflow-tensorboard    1.5.0                     <pip>
testpath                  0.3.1            py36h2698cfe_0
tornado                   4.5.3                    py36_0
traitlets                 4.3.2            py36h096827d_0
typing                    3.6.4                    py36_0
untangle                  1.1.1                     <pip>
urllib3                   1.22             py36h276f60a_0
vc                        14                   h0510ff6_3
vs2015_runtime            14.0.25123                    3
wcwidth                   0.1.7            py36h3d5aa90_0
webencodings              0.5.1            py36h67c50ae_1
werkzeug                  0.14.1                   py36_0
wheel                     0.30.0           py36h6c3ec14_1
win_inet_pton             1.0.1            py36he67d7fd_1
wincertstore              0.2              py36h7fe50ca_0
wrapt                     1.10.11          py36he5f5981_0
zlib                      1.2.11               h8395fce_2
```

**Comment by CAM-Gerlach** [comment]
@VanitarNordic Thanks. Can you try downgrading to 1.6.6 (with ``conda install sphinx=1.6.6`` and see if the error reoccurs? If not, can you try to locate the version that stops causing the problem? If you can't find it, did you update anything else at the same time you updated Sphinx and first saw the problem? 

The environment information is nice, but probably (hopefully) not necessary. What I asked for was what was stated in the new issue template that was apparently deleted/not followed, specifically your Qt and PyQt versions (from Help > About) and main dependencies (From Help > Dependencies > Copy to clipboard).

**Comment by VanitarNordic** [comment]
Dependencies:

```
IPython >=4.0    :  6.2.1 (OK)
cython >=0.21    :  None (NOK)
jedi >=0.9.0     :  0.11.1 (OK)
nbconvert >=4.0  :  5.3.1 (OK)
numpy >=1.7      :  1.14.1 (OK)
pandas >=0.13.1  :  0.22.0 (OK)
pycodestyle >=2.3:  2.3.1 (OK)
pyflakes >=0.6.0 :  1.6.0 (OK)
pygments >=2.0   :  2.2.0 (OK)
pylint >=0.25    :  1.8.2 (OK)
qtconsole >=4.2.0:  4.3.1 (OK)
rope >=0.9.4     :  0.9.4-1 (OK)
sphinx >=0.6.6   :  1.7.1 (OK)
sympy >=0.7.3    :  None (NOK)
```
QT and PyQT: 
![2018-03-02_3-33-35](https://user-images.githubusercontent.com/18719591/36876500-8f765000-1dca-11e8-9050-dd0bf6baf4e2.jpg)

> Can you try downgrading to 1.6.6 (with conda install sphinx=1.6.6 and see if the error reoccurs?

I'll try downgrading a.s.a.p. I remember that I updated some other components by Conda commands either, but I don't remember what they were. For now before testing the downgrading, I opened closed/re-opened the Spyder and the error does not appear by the same procedure. Therefore we can say it does not always appear, but when it appears, it is consistent.

**Comment by CAM-Gerlach** [comment]
Thanks. Let us know what you find.

> Therefore we can say it does not always appear, but when it appears, it is consistent.

So again, just to be clear: This consistently you speak of is within the same Spyder session?

**Comment by VanitarNordic** [comment]
> So again, just to be clear: This consistently you speak of is within the same Spyder session?

Yes, I always open just one Spyder session. When it appears like few minutes ago, it is consistent and reproducible in that session. I closed and opened the Spyder again. it did not appear so far nor producible.

**Comment by CAM-Gerlach** [comment]
Hmm, interesting. So it has happened over multiple previous sessions? Any pattern to that? Also, anything different in your current session—project location/working directory, files open, etc?

**Comment by VanitarNordic** [comment]
@CAM-Gerlach 

Aha, now it appeared again in the same session which I told you `it has not appeared so far`, 

Now, it can be produced by the same process. 

I don't do anything different to create just this error. For the pattern, it appears after I run a code. if I don't run a code it does not appear. The code is on Drive D and Anaconda is installed on Drive C as Windows. Everything including the code is identical before I upgrade the Sphinx.

**Comment by CAM-Gerlach** [comment]
> The code is on Drive D and Anaconda is installed on Drive C as Windows

Well there we go, which is why in my original followup, I asked

>  Is any of your code on another drive?

With that in hand, I was able to immediately reproduce it after switching my working directory (top right) to another drive, and then went to "Show tutorial" as you mentioned. Therefore, it seems likely the error has to do with use of a relative path (i.e. relative to the working dir) that attempts to access a location on another windows drive somewhere in the Sphinx call. From there, it shouldn't be too difficult for someone to trace and fix. I appreciate your help!

**Comment by CAM-Gerlach** [comment]
I can confirm that all sphinx help is in fact broken (not showing the tutorial, etc. at all, and only showing plain text of docstrings for Sphinx 1.7.0+ on Windows under the above stated conditions. I can confirm Sphinx <= 1.6.6 is unaffected, so the location for now is to downgrade to that with ``conda install sphinx=1.6.6`` from the Anaconda prompt (or just don't use a working directory different from your Python/Anaconda install).

**Comment by VanitarNordic** [comment]
@CAM-Gerlach 

I had to leave the discussion because it was midnight in my time. No part of my code is located on another Drive. All parts are in a single file on Drive D, which Anaconda and Windows are on Drive C. It is the identical situation as it was before upgrading the Sphinx. Now you reproduced it. Thank you for your follow up and patience.

**Comment by CAM-Gerlach** [comment]
Sorry for the confusion. What I meant by

> Is any of your code on anther drive?

is another drive *relative to Anaconda*; I guess I assumed that would be the most straightforward interpretation (or at least a potential interpretation, such that it would have solicited a response if anything deviated from a "baseline" setup with all code and Anaconda on the same drive) but I guess should have been explicit specifying both. 

In any case, thanks to your help, the problem is finally isolated so it can be solved.

**Comment by RicardoNoguerales** [comment]
This error message is fired in **C:\Python27\Lib\site-packages\spyder\plugins\help.py** 
when an exception in method `run `of class `SphinxThread `is handled by the `try: except:` blocks

This handling prevents the full traceback been shown in Spyder's Internal terminal
If you deactivate the try-except, the traceback appears:
```
>>> Traceback (most recent call last):
  File "c:\python27\lib\site-packages\spyder\plugins\help.py", line 301, in run
    html_text = sphinxify(doc['docstring'], context)
  File "c:\python27\lib\site-packages\spyder\utils\help\sphinxify.py", line 215, in sphinxify
    sphinx_app.build(None, [rst_name])
  File "c:\python27\lib\site-packages\sphinx\application.py", line 344, in build
    'outdir': path.relpath(self.outdir),
  File "c:\python27\lib\ntpath.py", line 528, in relpath
    % (path_prefix, start_prefix))
ValueError: path is on drive c:, start on drive E:     

```
## The problem  explained

- Help section rendering uses Sphinx to convert rst text help to nice HTML.Sphinx logs this activity.
- Part of this logging is in `build `method of class `Sphinx`
  **c:\Python27\Lib\site-packages\sphinx\application.py** circa line 343
  ```
                logger.info(self.builder.epilog % {
                    'outdir': path.relpath(self.outdir),
                    'project': self.config.project
                })
  ```
   Calls `path.relpath` ( `os.path.relpath`, in Windows defined in **c:\python27\lib\ntpath.py** ) with a single   argument, some temporary dir usually in C:\
- Being called with a single argument, `relpath `uses second argument's default '.', the python current working dir, say `os.getcwd()`
- `relpath `fires a `ValueError `if called with paths from different drives.
  So if Spyder's working dir (python current working dir) is not in drive C: , the exception is fired

## Why once per session

This exception switches help display mode from _richt text_ to _plain text_

_plain text_ mode doesnt use Sphinx, so the problem dissapears unless you switch back to _richt text_ mode.

## SOLUTIONS

Being involved Sphinx, Spyder and ntpath.py, there are various options

- **Patch ntpath.py** 
 You could patch `relpath `in **c:\python27\lib\ntpath.py** , but this could be dangerous, other scripts can expect this exception been fired and handle it their own way.

- **Patch Sphinx** (my choose)
 **c:\Python27\Lib\site-packages\sphinx\application.py** this way
```
###                logger.info(self.builder.epilog % {
###                    'outdir': path.relpath(self.outdir),
###                    'project': self.config.project
###                })
                # PATCH----------------------------------------------
                try:
                    the_relpath = path.relpath(self.outdir)
                except ValueError:
                    the_relpath = self.outdir
                logger.info(self.builder.epilog % {
                    'outdir': the_relpath,
                    'project': self.config.project
                })
                # PACH----------------------------------------------
```

- **Override relpath in Spyder**
 In **C:\Python27\Lib\site-packages\spyder\utils\help\sphinxify.py**
 after line 32 , `from sphinx.application import Sphinx`
 insert this
```
#-----------------------------------------------------------#PATCH
real_osp_relpath = osp.relpath
def relpath_patched(path, start='.'):
    try:
        #print '**** relpath_patched OK ****',path
        return real_osp_relpath(path, start)
    except:
        #print '**** relpath_patched avoided exception ****',path
        return path
sphinx.application.os.path.relpath = relpath_patched
# or os.path.relpath = relpath_patched
#-----------------------------------------------------------#PATCH
```
 Warning: Seems to override `relpath `in all Spyder app modules, not only in sphinx.application

**Comment by ccordoba12** [comment]
@RicardoNoguerales, thanks for investigating the source of the problem and giving us a solution. We'll come up with a fix for this in our next release.

**Comment by dalthviz** [comment]
@ccordoba12 testing with `Sphinx 1.7.5` seems like this is no more an issue:

![prev](https://user-images.githubusercontent.com/16781833/40943473-dcc2341e-6816-11e8-847d-359263a92fe5.gif)

**Comment by ccordoba12** [comment]
Thanks @dalthviz! I figured it was no longer a problem since no one else commented here.

Closing then.

**Comment by koneshgit** [comment]
I had Sphinx 1.8.2, and the same problem. Downgraded to 1.7.5 and now it is working
